src/IO/S3/Client.cpp

```diff
@@ -51,11 +51,43 @@ namespace ErrorCodes
 namespace S3
 {
 
-Client::RetryStrategy::RetryStrategy(uint32_t maxRetries_, uint32_t scaleFactor_, uint32_t maxDelayMs_)
+Int64 calculateDelayOnRetries(const Int64 retries, const Int64 scaleFactor)
+{
+    return (1ul << std::min(retries, 31l)) * scaleFactor;
+}
+
+Client::RetryStrategy::RetryStrategy(uint32_t maxRetries_, ContextPtr context, uint32_t scaleFactor_, uint32_t maxDelayMs_)
     : maxRetries(maxRetries_)
     , scaleFactor(scaleFactor_)
     , maxDelayMs(maxDelayMs_)
 {
+    /// Calculating number of retries based on the `max_execution_time` and delays between retries
+    if (context != nullptr)
+    {
+        Int64 execution_time = context->getSettingsRef().max_execution_time.totalMilliseconds() / 2; /// Here we put total time/2 knowing that we will do retries
+                                                                                                    /// two times in the doRequest: first time in `getRegionForBucket()`
+                                                                                                    /// and the second time in `request_fn()` so that we can follow
+                                                                                                    /// `max_execution_time` correctly.
+        if (execution_time > 0)
+        {
+            Int64 retries = 0;
+            while (true)
+            {
+                retries++;
+                uint64_t calculated_delay = calculateDelayOnRetries(retries, scaleFactor);
+                if (calculated_delay > maxDelayMs)
+                {
+                    retries += execution_time / maxDelayMs;
+                    break;
+                }
+                execution_time -= calculated_delay;
+                if (execution_time < 0)
+                    break;
+            }
+            maxRetries = maxRetries < retries ? maxRetries : uint32_t(retries);
+        }
```

Comment:
I think we can write a formula here, instead of writing the loop.
First, we can calculate how many steps is required to get `maxDelayMs`: `maxDelayMs <= minDelayMs * (scaleFactor ^ steps) => steps = log_(scaleFactor)(maxDelayMs / minDelayMs) = ln(maxDelayMs / minDelayMs) / ln(scaleFactor)` (or just `log_2(maxDelayMs / minDelayMs)` if scaleFactor = 2).

Then, we can get the time required to get to `maxDelayMs`, its `t = minDelayMs * (1 - scaleFactor ^ steps) / (1 - scaleFactor)`. If `t > max_execution_time`, then `maxRetries = ln(max_execution_time / minDelayMs) / ln(scaleFactor)`. In other case, `maxRetries = steps + (max_execution_time - t) * maxDelayMs`

Reply:
(but please validate this)